### DISCUSSION TITLE
private type `FactoryWidgets` in public interface

### DISCUSSION THREAD

**Discussion by uvizhe** [Relm4/Relm4#159] [Q&A] [answered]
Hello,
I'm trying to split my code into modules and I am facing this issue:
When I move a factory with a relevant struct to a separate module I have to declare the struct public
```
pub struct MyFactoryData {
    ...
}

#[relm4::factory_prototype]
impl FactoryPrototype for MyFactoryData {
    type Widgets = FactoryWidgets;
    ...
}
```
so to be able to use it in `main.rs` in `AppModel` e.g.
```
struct AppModel {
    dataset: FactoryVec<MyFactoryData>
}
```
But this causes a compiler error
```error[E0446]: private type `FactoryWidgets` in public interface```

Can you suggest a solution? I am relatively new to Rust so it might be on a surface but I don't see it.

**Comment by AaronErhardt** [comment, marked as answer]
You should be able to use `#[relm4::factory_prototype(pub)]` to make the generated `FactoryWidgets` struct public.

Btw. this hasn't anything to do with Rust directly, but the macro generates the struct as not-public by default. And because the code generation is somewhat hidden, it's not easy to find this feature. It's documented [in the book](https://relm4.org/book/widget_macro_reference.html#public-widgets) quite at the end, at least for the widget macro.

**Comment by 0xfalafel** [comment]
Comming from Google, I had the same issue.

`#[relm4::factory_prototype(pub)]` doesn't exist anymore, and the link is dead.
But it's possible to use __`#[relm4::factory(pub)]`__ to obtain the __same effect__.
